inst/include/iterative_solver.h

```diff
@@ -0,0 +1,128 @@
+// taken from Edwin van Leeuwen at
+// https://gitlab.com/epidemics-r/code_snippets/feature/newton_solver/include/finalsize.hpp
+
+/// @brief Solve an equation using the Newton method with tolerance checking
+/// @param contact_matrix Social contact matrix. Entry \eqn{mm_{ij}} gives
+/// the average number of contacts in group \eqn{i} reported by participants
+/// in group \eqn{j}.
+/// @param demography_vector Demography vector. Entry \eqn{pp_{i}} gives
+/// proportion of total population in group \eqn{i}. Normalised if needed.
+/// @param susceptibility A matrix giving the susceptibility of individuals
+/// in demographic group \eqn{i} and risk group \eqn{j}.
+/// @param iterations Number of solver iterations. Defaults to 10,000.
+/// @param tolerance Solver error tolerance. Solving ends when the error
+/// drops below this tolerance. Defaults to set `1e-6`. Larger tolerance
+/// values are likely to lead to inaccurate final size estimates.
+/// @param step_rate The solver step rate. Defaults to 1.9 as a value found to
+/// work well.
+/// @param adapt_step Boolean, whether the solver step rate should be changed
+/// based on the solver error. Defaults to TRUE.
+/// @return A vector of final sizes.
+inline Eigen::ArrayXd solve_final_size_iterative(
+    const Eigen::MatrixXd &contact_matrix,
+    const Eigen::VectorXd &demography_vector,
+    const Eigen::VectorXd &susceptibility, const int iterations = 10000,
+    const double tolerance = 1e-6, double step_rate = 1.9,
+    const bool adapt_step = true) {
+  // count number of demography groups
+  int nDim = demography_vector.size();
+
+  Eigen::ArrayXi zeros(nDim);
+  zeros.fill(0);
+
+  Eigen::ArrayXd epi_final_size(nDim);  // prev in settings struct
+  epi_final_size.fill(0.5);
+
+  Eigen::MatrixXd contact_matrix_ = contact_matrix;
+  for (int i = 0; i < contact_matrix.rows(); ++i) {
+    // Check if value should be 0 for (limited) performance increase
+    if (demography_vector(i) == 0 || susceptibility(i) == 0 ||
+        contact_matrix.row(i).sum() == 0) {
+      zeros[i] = 1;
+      epi_final_size[i] = 0;
+    }
+    for (int j = 0; j < contact_matrix.cols(); ++j) {
+      if (zeros[j]) {
+        contact_matrix_(i, j) = 0;
+      } else {
+        // Scale contacts appropriately
+        contact_matrix_(i, j) = contact_matrix(i, j) * demography_vector(j);
+      }
+    }
+  }
+
+  Eigen::VectorXd epi_final_size_return(nDim);
+  // define functions to minimise error in final size estimate
+  auto f = [&contact_matrix_, &susceptibility, &zeros](
+               const Eigen::VectorXd &epi_final_size,
+               Eigen::VectorXd &epi_final_size_return) {
+    Eigen::VectorXd s = contact_matrix_ * (-epi_final_size);
+    for (int i = 0; i < contact_matrix_.rows(); ++i) {
+      if (zeros[i] == 1) {
+        epi_final_size_return[i] = 0;
+        continue;
+      }
+      epi_final_size_return[i] = 1;
+
+      epi_final_size_return(i) -= exp(susceptibility(i) * s(i));
+    }
+  };
```

Comment:
I've accepted this edit for the initial section, but I would prefer to not use r-value referencing and `std::move` semantics - I think modification by reference as in the original code is easier to understand. We also moved away from using `std::move` semantics since Cppcheck would throw an warning when accessing a `move`-d variable (to assign it to `epi_final_size_return`). This was implemented in PR #114.

The iteration step in this solver expects that `epi_final_size_return` is modified by reference (as it is not assigned), so `f()` must implicitly have a `void` return type. If I understand correctly, this isn't possible for a lambda hence the `auto` type.

Reply:
Fixed in https://github.com/epiverse-trace/finalsize/pull/161/commits/0faf35b5eca05ce2a5d899df17fdd5a11389ca7a and https://github.com/epiverse-trace/finalsize/pull/161/commits/89d5cc7eea21956d39a3491a784fd0bc8c198ac3